IPython/core/interactiveshell.py

```diff
@@ -2362,6 +2362,14 @@ def run_line_magic(self, magic_name: str, line, _stack_depth=1):
                 kwargs['local_ns'] = self.get_local_scope(stack_depth)
             with self.builtin_trap:
                 result = fn(*args, **kwargs)
+
+            # The code below prevents output from being displayed
+            # when using magic %time.
+            # Output from '%time foo();', for instance, would never
+            # be displayed.
+            if magic_name == "time" and len(magic_arg_s) > 0 and magic_arg_s[-1] == ";":
+                return None
```

Comment:
Comparing to the `quiet` implementation:

https://github.com/ipython/ipython/blob/28f76f8a2f37700f8d74e91c4a54929dc367000f/IPython/core/displayhook.py#L84-L103

this proposal does not seem to handle comments or flanking whitespaces; I know that comments in general are not handled well by magics right know (which is  a pity), but I think that there should be a test case for suffix space/tab etc to be a bit more permissive in interactive coding.

Reply:
Thanks for educating me. I am also not a big fan of comparing the magic name for a particular implementation directly in the code either. I simply took a look on a list of magics I found on the internet and thought that the code in the PR (excluding the magic name comparison) would be OK for every magic. I did not know magics are extensible. Since they are, the idea of creating a decorator makes sense.

As for your comment about comparing the last character against the code in `quiet()`: I am pretty sure that the semicolon at the end of the line is not detected by `quiet()` today for any implemented magic. Debugging the code, the command

`%time 1;  #comment `

creates the following statement that is parsed by `quiet()`:

`get_ipython().run_line_magic('time', '1;  #comment')`

As you see, ; is not the last token of the expression and therefore is not detected.

The fact that the semicolon never prevents the expression value from being calculated today when a magic is performed plus the fact that magics are extensible and can basically do anything, makes me think that using something like `quiet()` (which is based on Python syntax) is too restrictive. I guess that perhaps we can configure the sub-string as a parameter of the decorator.

Example:

The definition of the magic time would be:

```
    @skip_doctest
    @no_var_expand
    @needs_local_scope
    @line_cell_magic
    @prevent_output_printing('%%')  # <--- THIS WILL BE ADDED
    def time(self,line='', cell=None, local_ns=None):
    ...
```

`@prevent_output_printing` would remove the sub-string '%%' if it finds one at the end of the line, call the magic implementation (time in the example) and then force `None` as the return value if the sub-string was found.

A line like that:

`%time 1  # comment %%`

would not produce a '`1'` on the screen.

We could implement something like `quiet()` in the decorator too, but my concern is that this may be too restrictive.

What do you think? Please be aware that I am not a hard IPython user and I am completely sure you and other people in this forum have used the software much more than me.

Reply:
One more note: I am not sure I should convert this to draft even if it is not ready to be merged, since I am afraid if I draft it, not everybody will have access to this discussion. What should I do here, @krassowski? Thanks.

Reply:
Feel free to convert it to draft and back, it does not affect the visibility of discussion.

> magics are extensible and can basically do anything, makes me think that using something like quiet() (which is based on Python syntax) is too restrictive. I guess that perhaps we can configure the sub-string as a parameter of the decorator.

I don't think that we need to support arbitrary suffixes; I am not against this in principle, but this may introduce interaction with other parsing bits, so I would suggest to postpone it possibly for another PR.

Reply:
New attempt. Please take a look. I don't know why, but darker complained about the single quotes in _ipython_magic_no_var_expand, which I changed to double quotes. Thanks.